# 결측치 처리

In [1]:
# 라이브러리 불러오기
import numpy as np
import pandas as pd
import seaborn as sns #시각화 라이브러리

###### titanic 데이터셋 가져오기

In [2]:
df = sns.load_dataset('titanic')
df.head()

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
0         0       3    male  22.0      1      0   7.2500        S  Third   
1         1       1  female  38.0      1      0  71.2833        C  First   
2         1       3  female  26.0      0      0   7.9250        S  Third   
3         1       1  female  35.0      1      0  53.1000        S  First   
4         0       3    male  35.0      0      0   8.0500        S  Third   

     who  adult_male deck  embark_town alive  alone  
0    man        True  NaN  Southampton    no  False  
1  woman       False    C    Cherbourg   yes  False  
2  woman       False  NaN  Southampton   yes   True  
3  woman       False    C  Southampton   yes  False  
4    man        True  NaN  Southampton    no   True

In [3]:
df.shape

(891, 15)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype   
---  ------       --------------  -----   
 0   survived     891 non-null    int64   
 1   pclass       891 non-null    int64   
 2   sex          891 non-null    object  
 3   age          714 non-null    float64 
 4   sibsp        891 non-null    int64   
 5   parch        891 non-null    int64   
 6   fare         891 non-null    float64 
 7   embarked     889 non-null    object  
 8   class        891 non-null    category
 9   who          891 non-null    object  
 10  adult_male   891 non-null    bool    
 11  deck         203 non-null    category
 12  embark_town  889 non-null    object  
 13  alive        891 non-null    object  
 14  alone        891 non-null    bool    
dtypes: bool(2), category(2), float64(2), int64(4), object(5)
memory usage: 80.6+ KB


## 결측치(NaN, null) 확인

###### deck 열의 NaN 개수 계산하기

In [5]:
nan_deck = df['deck'].value_counts(dropna=False)
nan_deck

NaN    688
C       59
B       47
D       33
E       32
A       15
F       13
G        4
Name: deck, dtype: int64

###### isnull() 메서드로 누락 데이터 찾기

In [6]:
df.isnull()

survived  pclass    sex    age  sibsp  parch   fare  embarked  class  \
0       False   False  False  False  False  False  False     False  False   
1       False   False  False  False  False  False  False     False  False   
2       False   False  False  False  False  False  False     False  False   
3       False   False  False  False  False  False  False     False  False   
4       False   False  False  False  False  False  False     False  False   
..        ...     ...    ...    ...    ...    ...    ...       ...    ...   
886     False   False  False  False  False  False  False     False  False   
887     False   False  False  False  False  False  False     False  False   
888     False   False  False   True  False  False  False     False  False   
889     False   False  False  False  False  False  False     False  False   
890     False   False  False  False  False  False  False     False  False   

       who  adult_male   deck  embark_town  alive  alone  
0    False       False   True        False  False  False  
1    False       False  False        False  False  False  
2    False       False   True        False  False  False  
3    False       False  False        False  False  False  
4    False       False   True        False  False  False  
..     ...         ...    ...          ...    ...    ...  
886  False       False   True        False  False  False  
887  False       False  False        False  False  False  
888  False       False   True        False  False  False  
889  False       False  False        False  False  False  
890  False       False   True        False  False  False  

[891 rows x 15 columns]

###### notnull() 메서드로 누락 데이터 찾기

In [7]:
df.head().notnull()

survived  pclass   sex   age  sibsp  parch  fare  embarked  class   who  \
0      True    True  True  True   True   True  True      True   True  True   
1      True    True  True  True   True   True  True      True   True  True   
2      True    True  True  True   True   True  True      True   True  True   
3      True    True  True  True   True   True  True      True   True  True   
4      True    True  True  True   True   True  True      True   True  True   

   adult_male   deck  embark_town  alive  alone  
0        True  False         True   True   True  
1        True   True         True   True   True  
2        True  False         True   True   True  
3        True   True         True   True   True  
4        True  False         True   True   True

In [8]:
# isnull() 메서드로 누락 데이터 개수 구하기

In [9]:
df.head().isnull().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
deck           3
embark_town    0
alive          0
alone          0
dtype: int64

## 결측치 처리: 삭제 및 채우기

###### NaN 값이 500개 이상인 열을 모두 삭제 - deck 열(891개 중 688개의 NaN 값)

In [10]:
df_thresh = df.dropna(axis=1, thresh=500)  
df_thresh.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'embark_town', 'alive',
       'alone'],
      dtype='object')

###### age 열에 나이 데이터가 없는 모든 행을 삭제 - age 열(891개 중 177개의 NaN 값)

In [11]:
df_age = df.dropna(subset=['age'], how='any', axis=0)  
len(df_age)

714

###### df 데이터프레임에서 어떤 NaN이 포함이 되어 있으면 행을 삭제

In [12]:
df.dropna(how='any')

survived  pclass     sex   age  sibsp  parch     fare embarked  class  \
1           1       1  female  38.0      1      0  71.2833        C  First   
3           1       1  female  35.0      1      0  53.1000        S  First   
6           0       1    male  54.0      0      0  51.8625        S  First   
10          1       3  female   4.0      1      1  16.7000        S  Third   
11          1       1  female  58.0      0      0  26.5500        S  First   
..        ...     ...     ...   ...    ...    ...      ...      ...    ...   
871         1       1  female  47.0      1      1  52.5542        S  First   
872         0       1    male  33.0      0      0   5.0000        S  First   
879         1       1  female  56.0      0      1  83.1583        C  First   
887         1       1  female  19.0      0      0  30.0000        S  First   
889         1       1    male  26.0      0      0  30.0000        C  First   

       who  adult_male deck  embark_town alive  alone  
1    woman       False    C    Cherbourg   yes  False  
3    woman       False    C  Southampton   yes  False  
6      man        True    E  Southampton    no   True  
10   child       False    G  Southampton   yes  False  
11   woman       False    C  Southampton   yes   True  
..     ...         ...  ...          ...   ...    ...  
871  woman       False    D  Southampton   yes  False  
872    man        True    B  Southampton    no   True  
879  woman       False    C    Cherbourg   yes  False  
887  woman       False    B  Southampton   yes   True  
889    man        True    C    Cherbourg   yes   True  

[182 rows x 15 columns]

###### age 열의 첫 10개 데이터 출력 (5 행에 NaN 값)

In [13]:
df['age'].head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5     NaN
6    54.0
7     2.0
8    27.0
9    14.0
Name: age, dtype: float64

###### age 열의 NaN값을 다른 나이 데이터의 평균으로 변경하기

mean_age = df['age'].mean(axis=0)   # age 열의 평균 계산 (NaN 값 제외)
df['age'].fillna(mean_age)
df['age'].head(10)

###### embark_town 열의 829행의 NaN 데이터 출력

In [14]:
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
Name: embark_town, dtype: object

###### embark_town 열의 NaN값을 승선도시 중에서 가장 많이 출현한 값으로 치환하기

In [15]:
most_freq = df['embark_town'].value_counts().idxmax()
most_freq

'Southampton'

###### embark_town 열 829행의 NaN 데이터 출력 (NaN 값이 most_freq 값으로 대체)

In [16]:
df['embark_town'].fillna(most_freq)
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
Name: embark_town, dtype: object

###### embark_town 열의 829행의 NaN 데이터 출력

In [17]:
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829            NaN
Name: embark_town, dtype: object

In [18]:
# embark_town 열의 NaN값을 바로 앞에 있는 828행의 값으로 변경하기

In [19]:
df['embark_town'].fillna(method='ffill', inplace=True)
df['embark_town'][825:830]

825     Queenstown
826    Southampton
827      Cherbourg
828     Queenstown
829     Queenstown
Name: embark_town, dtype: object

## 중복값 처리하기 duplicated( )
- 동일한 관측값이 중복되는지 여부, 즉 행의 레코드가 중복되는지 여부를 확인하려면 duplicated() 메소드를 이용한다. 
- 전에 나온 행들과 비교하여 중복되는 행이면 True를 반환 하고, 처음 나오는 행에 대해서는 False를 반환한다.

###### 중복 데이터를 갖는 데이터프레임 만들기

In [20]:
df = pd.DataFrame({'c1':['a', 'a', 'b', 'a', 'b'],
                  'c2':[1, 1, 1, 2, 2],
                  'c3':[1, 1, 2, 2, 2]})
df

c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

###### 데이터프레임 전체 행 데이터 중에서 중복값 찾기

In [21]:
df.duplicated()

0    False
1     True
2    False
3    False
4    False
dtype: bool

###### 데이터프레임의 특정 열 데이터에서 중복값 찾기

In [22]:
df['c2'].duplicated()

0    False
1     True
2     True
3    False
4     True
Name: c2, dtype: bool

###### 데이터프레임에서 중복 행을 제거

In [23]:
df

c1  c2  c3
0  a   1   1
1  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

In [24]:
df.drop_duplicates()

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2
4  b   2   2

###### c2, c3열을 기준으로 중복 행을 제거

In [25]:
df.drop_duplicates(['c2', 'c3'])

c1  c2  c3
0  a   1   1
2  b   1   2
3  a   2   2